File: Code_MATLAB/NegativeLogLikelihood_EGARCH_asymmetric.m
function [negativeLL] = NegativeLogLikelihood_EGARCH_asymmetric(parameter_vector, returns)
    % Extract the parameters
    mu = parameter_vector(1,1);
    lambda = parameter_vector(2,1);
    phi = parameter_vector(3,1);
    kappa = parameter_vector(4,1);
    kappa_tilde = parameter_vector(5,1);
    nu = parameter_vector(6,1);

    % Run the Beta-t-EGARCH filter to get lambda and sigma
    [~, sigma, epsilon, ~, ~] = Filter_EGARCH_asymmetric(mu, lambda, phi, kappa, kappa_tilde, nu, returns);

    % Compute the implied residuals

    % Compute the log-likelihood of each observation
    LL = -log(sigma) + log(studentpdf(epsilon, nu));
    % Negative log likelihood sum (Equation 4.2 in case study)
    negativeLL = -sum(LL);

    % Close the function
end

File: Code_Python/negative_log_likelihood_egarch_asymmetric.py
# negative_log_likelihood_egarch_asymmetric.py
import numpy as np
from scipy.special import gamma
from filter_egarch_asymmetric import FilterEGARCHAsymmetric

class NegativeLogLikelihoodEGARCHAsymmetric:
    def student_t_pdf(self, x, nu):
        const = gamma((nu + 1) / 2) / (np.sqrt(np.pi * (nu - 2)) * gamma(nu / 2))
        return const * (1 + x**2 / (nu - 2))**(-(nu + 1) / 2)

    def __call__(self, parameter_vector, returns):
        mu, initial_lambda, phi, kappa, kappa_tilde, nu = parameter_vector

        filter_instance = FilterEGARCHAsymmetric()
        _, sigma, epsilon, _, _ = filter_instance(mu, initial_lambda, phi, kappa, kappa_tilde, nu, returns)

        ll = -np.log(sigma) + np.log(self.student_t_pdf(epsilon, nu))
        negative_ll = -np.sum(ll)

        return negative_ll
